### PR TITLE
ENH: support complex dtype in LSMR

### PR DESCRIPTION
**Pull request by kohr-h**
Simple fix to support solving problems with complex matrix and/or RHS. I was checking my options for inverting NUFFTs. LSQR worked, LSMR not.

~Where do the tests go?~ Found them

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
I checked the Azure cov report & everything added / adjusted here is covered by tests, which is a good sign.

I'll likely let a resident expert comment before merging if possible.

File: scipy/sparse/linalg/isolve/lsmr.py [hunk 1]
```diff
@@ -221,6 +221,8 @@ def lsmr(A, b, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
     if maxiter is None:
         maxiter = minDim
 
+    dtype = result_type(b, float)
```
Comment (pv) [inline scipy/sparse/linalg/isolve/lsmr.py]: dtype of A probably also should be checked?

Does LSMR work for complex A?
If yes, the docstring should be updated for the meaning of `^T` (does it become `^H`?).

**Review by pv — CHANGES_REQUESTED** [review]

File: scipy/sparse/linalg/isolve/lsmr.py (hunk 1 above)
Comment (kohr-h) [reply scipy/sparse/linalg/isolve/lsmr.py]: True, and actually one of the tests already covered that case. I've extended the dtype to also consider `x0` if given.

More tests included now.

File: scipy/sparse/linalg/isolve/lsmr.py [hunk 2]
```diff
@@ -221,6 +221,11 @@ def lsmr(A, b, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
     if maxiter is None:
         maxiter = minDim
 
+    if x0 is None:
+        dtype = result_type(A, b, float)
+    else:
+        dtype = result_type(A, b, x0, float)
```
Comment (kohr-h) [inline scipy/sparse/linalg/isolve/lsmr.py]: Actually one could always do this, because `None` behaves the same as `float`. But maybe that's confusing?

Comment (pv) [reply scipy/sparse/linalg/isolve/lsmr.py]: Better to be explicit here, I think.

**Comment by kohr-h** [comment]
Circle fails due to missing pybind11 during build, Travis fails since it uses a 1.17 dev build of NumPy which seems to add some deprecation warning related to broadcasting.

**Pull request merged**